### DISCUSSION TITLE
「作成した下書きを送信」エラー

### DISCUSSION THREAD

**Discussion by kawano164** [ttsukagoshi/mail-merge-for-gmail#279] [Q&A]
初めまして、差込メールのソフトを探していてたどり着きました。
大変便利に使わせていただいています。
ありがとうございます。

Githubを使うのが初めてなので、投稿の仕方など不適切でしたら誠に申し訳ありません。

送る件数が数件だとエラーが起きないのですが、30件程度にすると下記のようにエラーメッセージが表示されます。

パターン①
——————————————————
1. 30件程度の宛先をGoogleスプレッドシートに記載
2. 「下書きを作成」ボタンを押す→正常に下書きが生成される
3. 「作成した下書きを送信」ボタンを押す→下記のエラーメッセージが表示される
at sendDrafts (group-merge:1377:15)
    at sendCreatedDrafts (group-merge:1312:28)
    at _GS_INTERNAL_top_function_call_.gs:1:8
4. その後、何回か「作成した下書きを送信」を押していると、同じエラーメッセージが表示されるが、10〜20件ずつ程度下書きが送信される
5. 何件かは送信されずに下書きとして残る（Gmailの通常の送信ボタンを押して送信は可能）

また、上記のメッセージとの前後関係は把握してないのですが、3〜4の間に、自分のメールボックスに下記のエラーメッセージも届きます。
- - - - - - - - - - - -
予期しないエラー：
Exception: 無効なメール: -
at mailMerge (group-merge:1135:28)
at postProcessMailMerge (group-merge:1302:3)
at __GS_INTERNAL_top_function_call__.gs:1:8
- - - - - - - - - - - -

パターン②
——————————————————
1〜2はパターン①と同一
3. 作成した下書きを送信ボタンを押すと下記のメールが連続で届く
- - - - - - - - - - - -
予期しないエラー：
Exception: 無効なメール: -
at mailMerge (group-merge:1135:28)
at postProcessMailMerge (group-merge:1302:3)
at __GS_INTERNAL_top_function_call__.gs:1:8
- - - - - - - - - - - -

- - - - - - - - - - - -
予期しないエラー：
Exception: Gmail の操作を行うことはできません。
at sendDrafts (group-merge:1377:15)
at postProcessSendDrafts (group-merge:1511:3)
at __GS_INTERNAL_top_function_call__.gs:1:8
- - - - - - - - - - - -

Group Mergeで送信されなかった下書きも直接送信ボタンを押せば送信できるので、今のところ実務上は大きな問題にはなっていません。

ただ、エラーの原因や解決方法など、もしありましたらご教授いただければ幸いです。

**Comment by ttsukagoshi** [comment]
@kawano164 ご連絡ありがとうございます。開発者です。
いくつかのエラーが重なっているようにお見受けします。エラーメッセージ `Exception: 無効なメール` や `Exception: Gmail の操作を行うことはできません。` はいずれもかなり一般的なエラーらしく...原因を精査させていただきたいのですが、もう少し詳しく、以下の点を共有いただけますでしょうか：

## 1) パターン①、②とも下書きが**全て**正常に生成されているか、今一度ご確認いただけますか？
「2.「下書きを作成」ボタンを押す→正常に下書きが生成される」という段階で、スプレッドシートに記載した「30件程度の宛先」が全て下書きとして生成されているか、ご確認いただけますでしょうか。

次の1, 2をいずれも満たしていれば大丈夫です：
1. 「下書きを作成」ボタンを押した後、アドオン画面内で「完了」のメッセージが表示される：
![下書き作成後の「完了」メッセージが表示されたアドオン画面のスクリーンショット](https://user-images.githubusercontent.com/55706659/214466717-12fbc1f7-36ba-4f53-b3fd-8c6e8a1b8fd7.png)
2. 完了メッセージの「○○通の下書きが作成されました」の下書きメール件数が、スプレッドシートの宛先数と一致する

→この質問の意図は、下書き作成のプロセスが完了する前に「作成した下書きを送信」ボタンを押されている可能性を考えてのものです。というのも、共有いただいたエラーメッセージのうち

> 予期しないエラー：
> Exception: 無効なメール: -
> at mailMerge (group-merge:1135:28)
> at postProcessMailMerge (group-merge:1302:3)
> at GS_INTERNAL_top_function_call.gs:1:8

については `postProcessMailMerge` という処理の最中に発生したエラーであることを示しています。
これは、Googleが定める30秒という実行時間制限では差し込みメールの下書き作成が完了しなかったときに、内部的に下書き作成処理を引き継いで実行する別処理の名前です。少なくともこのエラーが発生した時点では、下書きメールの作成が完了していなかったことを示しているのです。

下書きの作成が完了していない中で「作成した下書きを送信」ボタンを押すと、予期しないエラーの原因となる可能性があります（処理がバッティングしないよう予防する仕組みは、現時点では入っていないです）。「下書きを作成」が別処理（`postProcessMailMerge`）に引き継がれた際には、アドオン画面で以下↓のようなメッセージが表示されるとともに、別処理完了後＝下書きが全て作成された際にはアドオン実行アカウント宛に完了のメール通知が飛びます。
<img width="296" alt="image" src="https://user-images.githubusercontent.com/55706659/214469730-fa04f611-53a6-4a60-8722-736ab6cc38ef.png">
もしこれが原因だった場合は、下書き作成が完了するまでお待ちいただいてから「作成した下書きを送信」ボタンを押すと、エラーは出なくなるかもしれません。

## 2) エラーメッセージの全体像を拝見できますか？
パターン①で書いていただいたエラーメッセージについて：

> 3.「作成した下書きを送信」ボタンを押す→下記のエラーメッセージが表示される
> at sendDrafts (group-merge:1377:15)
> at sendCreatedDrafts (group-merge:1312:28)
> at GS_INTERNAL_top_function_call.gs:1:8

→エラーメッセージの1行目に、他のメッセージ同様、一言あったかと思うのですが、そちらも共有いただけますか？

パターン②で入れていただいた

> Exception: Gmail の操作を行うことはできません。
> at sendDrafts (group-merge:1377:15)
> at postProcessSendDrafts (group-merge:1511:3)
> at GS_INTERNAL_top_function_call.gs:1:8

と同様に `Exception: Gmail の操作を行うことはできません。` なのではと拝察するのですが、念のため、です。

ご協力のほど、よろしくお願いします。

**Reply by kawano164** [reply]
@ttsukagoshi お忙しい中丁寧なコメントいただきまして誠にありがとうございます。
ちょっとお時間かかってしまうかもしれませんが、いただいた点を必ず確認してまたこちらにご連絡させていただきます。
引き続きどうぞよろしくお願いいたします。

**Reply by kawano164** [reply]
ご返事大変遅くなりました。
先ほど、久しぶりにGroup Mergeを使わせていただく機会がありまして（大変重宝しております。本当にありがとうございます）、同じエラーが発生しましたので、改めてご報告させていただきます。

1)下書きは正常に作成されているようです。
件数が多い場合Group Mergeの完了画面では21件の下書きを作成しました。残りが完了したらメールします云々と表示され、下記の文章のメールが数分後に届きます。
- - - - - - - - - - - -
完了：64通の下書きが作成されました。差し込み済みのメールが表示されない場合、
Gmailの下書き画面を更新してください。
- - - - - - - - - - - -
ここでの64通というのは、元になったスプレッドシートの件数と同一です。

2)エラーの全文
以下がエラーの全文です。
- - - - - - - - - - - -
予期しないエラー：
Exception: Gmail の操作を行うことはできません。
at sendDrafts (group-merge:1377:15)
at sendCreatedDrafts (group-merge:1312:28)
at _GS_INTERNAL_top_function_call_.gs:1:8
- - - - - - - - - - - -
